File: src/main/java/com/nadmm/airports/wx/WxTextFragmentBase.java
/*
 * FlightIntel for Pilots
 *
 * Copyright 2012-2017 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.nadmm.airports.wx;

import android.content.BroadcastReceiver;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.nadmm.airports.R;
import com.nadmm.airports.utils.TextFileViewActivity;

public abstract class WxTextFragmentBase extends WxFragmentBase {

    private final String mAction;
    private final String[] mWxAreaCodes;
    private final String[] mWxAreaNames;

    private View mPendingRow;

    public WxTextFragmentBase( String action, String[] areaCodes, String[] areaNames ) {
        mAction = action;
        mWxAreaCodes = areaCodes;
        mWxAreaNames = areaNames;
    }

    @Override
    public void onCreate( Bundle savedInstanceState ) {
        super.onCreate( savedInstanceState );

        setupBroadcastFilter( mAction );
    }

    @Override
    public View onCreateView( LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState ) {
        View view = inflate( R.layout.wx_map_detail_view );
        return createContentView( view );
    }

    @Override
    public void onViewCreated( View view, Bundle savedInstanceState ) {
        super.onViewCreated( view, savedInstanceState );

        TextView tv = (TextView) view.findViewById( R.id.wx_map_label );
        tv.setText( "Select Area" );
        tv.setVisibility( View.VISIBLE );

        OnClickListener listener = v -> {
            if ( mPendingRow == null ) {
                mPendingRow = v;
                String code = (String) v.getTag();
                requestWxText( code );
            }
        };

        LinearLayout layout = view.findViewById( R.id.wx_map_layout );
        for ( int i = 0; i < mWxAreaNames.length; ++i ) {
            View row = addWxRow( layout, mWxAreaNames[ i ], mWxAreaCodes[ i ] );
            row.setOnClickListener( listener );
        }

        setFragmentContentShown( true );
    }

    @Override
    public void onResume() {
        super.onResume();

        mPendingRow = null;
    }

    @Override
    protected void handleBroadcast( Intent intent ) {
        if ( mPendingRow != null ) {
            TextView tv = mPendingRow.findViewById( R.id.text );
            String label = tv.getText().toString();
            String path = intent.getStringExtra( NoaaService.RESULT );
            Intent viewer = new Intent( getActivity(), TextFileViewActivity.class );
            viewer.putExtra( TextFileViewActivity.FILE_PATH, path );
            viewer.putExtra( TextFileViewActivity.TITLE_TEXT, getTitle() );
            viewer.putExtra( TextFileViewActivity.LABEL_TEXT, label );
            getActivity().startActivity( viewer );
            setProgressBarVisible( false );
            mPendingRow = null;
        }
    }

    private void requestWxText( String code ) {
        setProgressBarVisible( true );

        Intent service = getServiceIntent();
        service.setAction( mAction );
        service.putExtra( NoaaService.STATION_ID, code );
        getActivity().startService( service );

        getActivityBase().faLogViewItem( getProduct(), code );
    }

    private void setProgressBarVisible( boolean visible ) {
        if ( mPendingRow != null ) {
            View view = mPendingRow.findViewById( R.id.progress );
            view.setVisibility( visible? View.VISIBLE : View.INVISIBLE );
        }
    }

    protected abstract String getTitle();

    protected abstract Intent getServiceIntent();

}

File: src/main/java/com/nadmm/airports/wx/WxTextFragmentBase.kt
/*
 * FlightIntel for Pilots
 *
 * Copyright 2012-2025 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nadmm.airports.wx

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import com.nadmm.airports.R
import com.nadmm.airports.databinding.ListItemText1Binding
import com.nadmm.airports.databinding.WxMapDetailViewBinding
import com.nadmm.airports.utils.TextFileViewActivity

abstract class WxTextFragmentBase(
    action: String,
    private val wxAreas: Map<String, String>,
    private val wxTypes: Map<String, String> = mapOf()
) : WxFragmentBase(action) {
    private var selectedRow: View? = null
    private var _binding: WxMapDetailViewBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = WxMapDetailViewBinding.inflate(inflater)
        return createContentView(binding.root)
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.apply {
            wxMapLabel.text = "Select Area"
            wxMapLabel.visibility = View.VISIBLE

            val listener = View.OnClickListener { v: View? ->
                if (selectedRow == null && v != null) {
                    selectedRow = v
                    val code = v.tag as String
                    requestWxText(code)
                }
            }

            for (area in wxAreas) {
                val row = addWxRow(wxMapLayout, area.value, area.key)
                row.setOnClickListener(listener)
            }

            if (helpText.isNotEmpty()) {
                wxHelpText.text = helpText
                wxHelpText.visibility = View.VISIBLE
            }

            if (wxTypes.isNotEmpty()) {
                wxMapTypeLabel.visibility = View.VISIBLE
                wxMapTypeLayout.visibility = View.VISIBLE
                val types = wxTypes.toSortedMap().values.toList()
                val adapter = ArrayAdapter(
                    requireActivity(),
                    R.layout.list_item, types)
                val textView = getAutoCompleteTextView(wxMapType)
                textView?.setAdapter(adapter)
                textView?.setText(types.first(), false)
            }
        }

        setFragmentContentShown(true)
    }

    override fun onResume() {
        super.onResume()

        selectedRow = null
    }

    override fun processResult(result: Bundle) {
        val path = result.getString(NoaaService.RESULT)
        if (!path.isNullOrEmpty()) {
            val item = ListItemText1Binding.bind(selectedRow!!)
            val label = item.text.text.toString()
            val viewer = Intent(activity, TextFileViewActivity::class.java)
                .apply {
                    putExtra(TextFileViewActivity.FILE_PATH, path)
                    putExtra(TextFileViewActivity.TITLE_TEXT, title)
                    putExtra(TextFileViewActivity.LABEL_TEXT, label)
                }
            activity?.startActivity(viewer)
        }
        setSpinnerVisible(false)
        selectedRow = null
    }

    private fun requestWxText(code: String?) {
        setSpinnerVisible(true)
        val service = serviceIntent.apply {
            setAction(this@WxTextFragmentBase.action)
            putExtra(NoaaService.TEXT_CODE, code)
            if (wxTypes.isNotEmpty()) {
                val text = getSelectedItemText(binding.wxMapType)
                val type = wxTypes.entries.first { it.value == text }.key
                putExtra(NoaaService.TEXT_TYPE, type)
            }
        }
        activity?.startService(service)
    }

    private fun setSpinnerVisible(visible: Boolean) {
        if (selectedRow != null) {
            val view = selectedRow!!.findViewById<View>(R.id.progress)
            view.visibility = if (visible) View.VISIBLE else View.INVISIBLE
        }
    }

    protected abstract val title: String
    protected abstract val helpText: String
    protected abstract val serviceIntent: Intent
}
